src/dma/traits.rs

```diff
@@ -0,0 +1,1394 @@
+    fn address(&self) -> u32;
+}
+
+// Convenience macro for implementing addresses on peripherals
+macro_rules! address {
+    ($(($peripheral:ty, $register:ident, $size: ty)),+ $(,)*) => {
+        $(
+            impl Sealed for $peripheral {}
+
+            impl PeriAddress for $peripheral {
+                #[inline]
+                fn address(&self) -> u32 {
+                    &self.$register as *const _ as u32
+                }
+
+                type MemSize = $size;
+            }
+        )+
+    };
+}
+
+impl Sealed for DMA1 {}
+impl Sealed for DMA2 {}
+
+#[cfg(not(any(
+    feature = "stm32f411",
+    feature = "stm32f413",
+    feature = "stm32f423",
+    feature = "stm32f410"
+)))]
+/// Type alias to a DMA RegisterBlock.
+pub type DMARegisterBlock = pac::dma2::RegisterBlock;
+
+#[cfg(any(
+    feature = "stm32f411",
+    feature = "stm32f413",
+    feature = "stm32f423",
+    feature = "stm32f410"
+))]
+/// Type alias to a DMA RegisterBlock.
+pub type DMARegisterBlock = pac::dma1::RegisterBlock;
+
+/// Trait that represents an instance of a DMA peripheral.
+pub trait Instance: Deref<Target = DMARegisterBlock> + Sealed {
+    /// Gives a pointer to the RegisterBlock.
+    fn ptr() -> *const DMARegisterBlock;
+}
+
+impl Instance for DMA1 {
+    #[inline]
+    fn ptr() -> *const DMARegisterBlock {
+        DMA1::ptr()
+    }
+}
+
+impl Instance for DMA2 {
+    #[inline]
+    fn ptr() -> *const DMARegisterBlock {
+        DMA2::ptr()
+    }
+}
+
+/// Trait for peripheral's clock enabling.
+pub trait RccEnable: Sealed {
+    /// Enable the peripheral's clock in the RCC.
+    fn rcc_enable(&self);
+}
+
+impl RccEnable for pac::DMA1 {
+    fn rcc_enable(&self) {
+        unsafe {
+            //NOTE(unsafe) this reference will only be used for atomic writes with no side effects
+            let rcc = &(*RCC::ptr());
+            // Enable and reset the timer peripheral
+            bb::set(&rcc.ahb1enr, 21);
+            bb::set(&rcc.ahb1rstr, 21);
+            bb::clear(&rcc.ahb1rstr, 21);
+        }
+    }
+}
+
+impl RccEnable for pac::DMA2 {
+    fn rcc_enable(&self) {
+        unsafe {
+            //NOTE(unsafe) this reference will only be used for atomic writes with no side effects
+            let rcc = &(*RCC::ptr());
+            // Enable and reset the timer peripheral
+            bb::set(&rcc.ahb1enr, 22);
+            bb::set(&rcc.ahb1rstr, 22);
+            bb::clear(&rcc.ahb1rstr, 22);
+        }
+    }
+}
+
+macro_rules! tim_channels {
+    ($(($name:ident, $register:ident)),+ $(,)*) => {
+        $(
+            /// Wrapper type that indicates which register of the contained timer to use for DMA.
+            pub struct $name<T> (T);
+
+            impl<T> Deref for $name<T> {
+                type Target = T;
+
+                #[inline]
+                fn deref(&self) -> &T {
+                    &self.0
+                }
+            }
+        )+
+    };
+}
```

Comment:
The type is used, but I think you're right about the `$register` field not doing anything. I didn't actually write that code, so I left it there, but maybe we should remove the `$register` part.

Reply:
I would remove it, a) because it is unused code and you wonder if there is a bug/typo somewhere here and b) because it makes readers think that "which register of the contained timer to use for DMA" is encoded in the `$register`, when it's actually encoded in the `$name`.

I guess `$register` may be there as additional documentation in the `tim_channels!` invocation, so you see that, e.g., the wrapper type `CCR1` refers the the register `ccr1`. But IMO that's not *too* hard to guess from the name already ;)